proposed/stream.md

```diff
@@ -0,0 +1,194 @@
+
+## 1. Specification
+
+`StreamInterface` exposes several methods that enable streams to be read
+from, written to, and traversed effectively.
+
+Streams expose their capabilities using three methods: `isReadable()`,
+`isWritable()`, and `isSeekable()`. These methods can be used by stream
+collaborators to determine if a stream is capable of their requirements.
+
+Each stream instance will have various capabilities: it can be read-only,
+write-only, or read-write. It can also allow arbitrary random access (seeking
+forwards or backwards to any location), or only sequential access (for
+example in the case of a socket, pipe, or callback-based stream).
+
+Finally, `StreamInterface` defines a `__toString()` method to simplify
+retrieving or emitting the entire body contents at once.
+
+## 2. Package
+
+The interfaces and classes described are provided as part of the
+[psr/stream](https://packagist.org/packages/psr/stream) package.
+
+## 3. Interface `Psr\Stream\StreamInterface`
+
+```php
+<?php
+
+namespace Psr\Stream;
+
+/**
+ * Describes a data stream.
+ *
+ * Typically, an instance will wrap a PHP stream; this interface provides
+ * a wrapper around the most common operations, including serialization of
+ * the entire stream to a string.
+ */
+interface StreamInterface
+{
+    /**
+     * Reads all data from the stream into a string, from the beginning to end.
+     *
+     * This method MUST attempt to seek to the beginning of the stream before
+     * reading data and read the stream until the end is reached.
+     *
+     * Warning: This could attempt to load a large amount of data into memory.
+     *
+     * This method MUST NOT raise an exception in order to conform with PHP's
+     * string casting operations.
+     *
+     * @see http://php.net/manual/en/language.oop5.magic.php#object.tostring
+     * @return string
+     */
+    public function __toString();
+
+    /**
+     * Closes the stream and any underlying resources.
+     *
+     * @return void
+     */
+    public function close();
+
+    /**
+     * Separates any underlying resources from the stream.
+     *
+     * After the stream has been detached, the stream is in an unusable state.
+     *
+     * @return resource|null Underlying PHP stream, if any
+     */
+    public function detach();
+
+    /**
+     * Get the size of the stream if known.
+     *
+     * @return int|null Returns the size in bytes if known, or null if unknown.
+     */
+    public function getSize();
+
+    /**
+     * Returns the current position of the file read/write pointer
+     *
+     * @return int Position of the file pointer
+     * @throws \RuntimeException on error.
+     */
+    public function tell();
+
+    /**
+     * Returns true if the stream is at the end of the stream.
+     *
+     * @return bool
+     */
+    public function eof();
+
+    /**
+     * Returns whether or not the stream is seekable.
+     *
+     * @return bool
+     */
+    public function isSeekable();
```

Comment:
I feel like it would be better to spin off the seekable methods into a separate interface so that it's easier to enforce on typehints. Ditto with readable and writable.

Reply:
i highly agree with @Chippiewill here, the API would look something in lines of : 

```php
interface StreamInterface {}

interface ReadStreamInterface extends
  StreamInterface {}
interface WriteStreamInterface extends
  StreamInterface {}

interface DuplexStreamInterface extends
  ReadStreamInterface, WriteStreamInterface {}

interface SeekableStreamInterface extends
  StreamInterface {}

interface SeekableReadStreamInterface extends
  SeekableStreamInterface, ReadStreamInterface {}
interface SeekableWriteStreamInterface extends
  SeekableStreamInterface, WriteStreamInterface {}

interface SeekableDuplexStreamInterface extends
  DuplexStreamInterface, SeekableReadStreamInterface, SeekableWriteStreamInterface {}
````